File: 1-D_Dynamic_Programming/Easy/LC_746-Min_Cost_Climbing_Stairs/LC_746-Min_Cost_Climbing_Stairs.py
class Solution:
    # T: O(n), M: O(1), where n is amount of steps
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        # Bottom-up approach; start from 3rd step
        for idx in range(2, len(cost)):
            cost[idx] += min(cost[idx-1], cost[idx-2])
        return min(cost[len(cost)-1], cost[len(cost)-2])

    '''
    # T: O(n), M: O(1), where n is amount of steps
    def minCostClimbingStairs(self, cost: list[int]) -> int:
        # Top-down approach; start from 3rd to last step
        for idx in range(len(cost)-3, -1, -1):
            cost[idx] += min(cost[idx+1], cost[idx+2])
        return min(cost[0], cost[1])
    '''

sol = Solution()

# Ex1
attempt = sol.minCostClimbingStairs([10, 15, 20])
assert attempt == 15, f"Expected 15, got {attempt}"
# Ex2
attempt = sol.minCostClimbingStairs([1, 100, 1, 1, 1, 100, 1, 1, 100, 1])
assert attempt == 6, f"Expected 6, got {attempt}"

File: 1-D_Dynamic_Programming/Easy/LC_746-Min_Cost_Climbing_Stairs/LC_746-Min_Cost_Climbing_Stairs.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // T: O(n), M: O(1), where n is amount of steps
    int minCostClimbingStairs(vector<int>& cost) {
        // Bottom-up approach; start from 3rd step
        for (int i = 2; i < cost.size(); ++i) {
            cost[i] += min(cost[i - 1], cost[i - 2]);
        }
        return min(cost[cost.size() - 1], cost[cost.size() - 2]);
    }

    /*
    // T: O(n), M: O(1), where n is amount of steps
    int minCostClimbingStairs(vector<int>& cost) {
        // Top-down approach; start from 3rd to last step
        for (int i = cost.size() - 3; i >= 0; --i) {
            cost[i] += min(cost[i + 1], cost[i + 2]);
        }
        return min(cost[0], cost[1]);
    }
    */
};

int main() {
    Solution sol;
    vector<int> cost;

    // Ex 1
    cost = {10, 15, 20};
    assert(sol.minCostClimbingStairs(cost) == 15);
    // Ex 2
    cost = {1, 100, 1, 1, 1, 100, 1, 1, 100, 1};
    assert(sol.minCostClimbingStairs(cost) == 6);
}